Views over graph and set variables

dimitri-justeau opened the pull request:
Fixes # .

Changes proposed in this PR:
- Implement and test observer pattern for set data structures. Observing set are storing their own data for a fast read access, but the observer pattern ensures that they are always reflecting a property from the sets they are observing. Observing sets that were implemented are : `SetDifference` (set difference of two sets), `SetUnion` (union of a list of sets), and `SetIntersection` (intersection of a list of sets).
- Implement and test arithmetic set views: `SetUnionView` (union of set variables), `SetDifferenceView` (set difference of two set variables), and `SetIntersectionView` (intersection of set variables). See `setUnionView`, `setDifferenceView`, and `setIntersectionView` in `IViewFactory`.
- Using observing set data structures, implement observing subgraph data structures (node-induced and edge-induced, see `GraphFactory`).
- Implement and test subgraph views (graph views over graph variables):
    - Node-induced graph views: from a set of nodes to exclude from the observed variable or from a set of nodes to intersect with the nodes of the observed variables. For directed and undirected graph vars (see `nodeInducedSubgraphView` method in `IViewFactory`).
    - Edge-induced graph views: from a set of edges to exclude from the observed variable or from a set of edges to intersect with the edges of the observed variables. For directed and undirected graph vars (see `edgeInducedSubgraphView` method in `IViewFactory`).
 
@chocoteam/core-developer

cprudhom commented:
There are two tests in failure: `SortedBitSetTest>SetTest.testReadOnlySet` and `RangeSetTest>SetTest.testReadOnlySet`.
Is there any quick fix? Otherwise, what is the difficulty behind?

dimitri-justeau commented on solver/src/main/java/org/chocosolver/solver/constraints/set/PropSumOfElements.java:
```diff
@@ -103,29 +103,41 @@ public void propagate(int evtmask) throws ContradictionException {
 			int j = iter.nextInt();
 			if (!set.getLB().contains(j)) {
-				if(min + get(j) > ub || max + get(j) < lb){
+				if (min + get(j) > ub) {
```
It has not really been removed, if you look at the following, I separated the different cases to accurately compute new bounds, and avoid using recursive calls to the propagator as it was in the previous version.

cprudhom replied:
You are right, I got confused with the part on the right

dimitri-justeau commented:
> There are two tests in failure: `SortedBitSetTest>SetTest.testReadOnlySet` and `RangeSetTest>SetTest.testReadOnlySet`.
> Is there any quick fix? Otherwise, what is the difficulty behind?

This was not a big issue, I just did not implement the observer pattern in IntIterableBitset and IntIterableRangeSet because they exposed other modifiers than `add`, `remove`, and `clear`. I also modified the ReadOnly set to make it forward modifications of the underlying set to observers of the ReadOnly version, by registering it as an observer of the underlying set. Thus instantiating ReadOnly sets with IntIterableBitset or IntIterableRangeset raised an `UnsupportedOperationException`.

I pushed a new commit where I implemented the observed pattern in these two IntIterable set classes. I tried to do it efficiently not spending too much time on it as IntIterableRangeset has a sophisticated behaviour. So the current solution works, but might not be the optimal one for sets observing IntIterableRangesets (but observer pattern is mainly useful for backtrackable structures used in setvars and graphvars so I don't believe this is really an issue). I also added a bit of Javadoc to new ISet methods related to this observer pattern.

dimitri-justeau reviewed (commented):
I enriched this PR by adding a new variable type: NON_INJECTIVE_VIEW, to avoid using such views in the default search strategy. Indeed, one instantiation of a non-injective view can correspond to different instantiations of observed variables, thus branching on such view can cause infinite loops or errors, and thus should not be the default behaviour of the solver. I also fixed edge-induced views that revealed bugs after the previous udpate. Everything should be ok now.
